src/Cryptol/REPL/Command.hs

```diff
@@ -391,32 +399,107 @@ dumpTestsCmd outFile str pos fnm =
-                  void (qcExpr qcMode doc texpr schema)
-
-qcCmd qcMode str pos fnm =
-  do expr <- replParseExpr str pos fnm
-     (_,texpr,schema) <- replCheckExpr expr
-     nd <- M.mctxNameDisp <$> getFocusedEnv
-     let doc = fixNameDisp nd (ppPrec 3 expr) -- function application has precedence 3
-     void (qcExpr qcMode doc texpr schema)
-
+qcCmd qcMode str pos fnm = do
+  -- (qcMde, doc, texpr, schema) <- case str of 
+  case str of
+    "" -> do
+      (xs,disp) <- getPropertyNames
+      let nameStr x = show (fixNameDisp disp (pp x))
+      if null xs
+      then do
+        rPutStrLn "There are no properties in scope."
+      else
+        forM_ xs $ \(x,d) -> do
+          let str = nameStr x
+          rPutStr $ "property " ++ str ++ " "
+          let texpr = T.EVar x
+          let schema = M.ifDeclSig d
+          nd <- M.mctxNameDisp <$> getFocusedEnv
+          let doc = fixNameDisp nd (pp texpr)
+          go doc Nothing texpr schema
+    _ -> do
+      expr <- replParseExpr str pos fnm
+      (_,texpr,schema) <- replCheckExpr expr
+      nd <- M.mctxNameDisp <$> getFocusedEnv
+      let doc = fixNameDisp nd (ppPrec 3 expr) -- function application has precedence 3
+      go doc (Just expr) texpr schema
+  where
+    go :: Doc -> Maybe (P.Expr P.PName) -> T.Expr -> T.Schema -> REPL ()
+    go doc mb_expr texpr schema = case qcMode of
+      QCExhaust -> do
+        testNum <- (toInteger :: Int -> Integer) <$> getKnownUser "tests"
+        (val, ty) <- replEvalCheckedExpr' texpr schema
+        void (qcExpr qcMode doc texpr schema testNum val ty Nothing)
+      QCRandom -> do
+        testNum <- (toInteger :: Int -> Integer) <$> getKnownUser "tests"
+        useLitSampling <- getKnownUser "literalSampling" :: REPL Bool
+        if  useLitSampling &&
+            not (null (T.sVars schema)) &&
+            all ((T.KNum ==) . T.tpKind) (T.sVars schema)
+          then do
+          case mb_expr of
+            Nothing -> do
+              (val, ty) <- replEvalCheckedExpr' texpr schema
+              void (qcExpr qcMode doc texpr schema testNum val ty Nothing)
+            Just expr -> do
+              -- do this many tests per sample, until out of tests
+              binSize <- (toInteger :: Int -> Integer) <$> getKnownUser "literalSamplingBinSize"
+              let bins = testNum `div` binSize
+              io (sampleLiterals schema (fromInteger bins)) >>= \case
+                Right samples_schemas -> do
+                  rPutStrLn "Using literal sampling."
+                  let -- this generates a new expression that has the appropriate
+                      -- type arguments given to it according to the sampling,
+                      -- which is then typechecked and evaluated again
+                      applySampleToTExpr :: Sample -> P.Expr P.PName -> P.Expr P.PName
+                      applySampleToTExpr sample e = P.EAppT e $ f <$> sample
+                        where
```

Comment:
This is a `where` inside a `let` inside a `case` inside an `if` inside a `where. I think we can find a way to break this up better.